substrate/runtime/staking/src/lib.rs

```diff
@@ -296,15 +327,70 @@ impl<T: Trait> Module<T> {
 	/// Retract the desire to stake for the transactor.
 	///
 	/// Effects will be felt at the beginning of the next era.
-	fn unstake(aux: &T::PublicAux) -> Result {
+	fn unstake(aux: &T::PublicAux, position: u32) -> Result {
+		let aux = aux.ref_into();
+		let position = position as usize;
 		let mut intentions = <Intentions<T>>::get();
-		let position = intentions.iter().position(|t| t == aux.ref_into()).ok_or("Cannot unstake if not already staked.")?;
+//		let position = intentions.iter().position(|t| t == aux.ref_into()).ok_or("Cannot unstake if not already staked.")?;
+		if intentions.get(position) != Some(aux) {
+			return Err("Invalid index")
+		}
 		intentions.swap_remove(position);
 		<Intentions<T>>::put(intentions);
 		<Bondage<T>>::insert(aux.ref_into(), Self::current_era() + Self::bonding_duration());
 		Ok(())
 	}
 
+	fn nominate(aux: &T::PublicAux, target: RawAddress<T::AccountId, T::AccountIndex>) -> Result {
+		let target = Self::lookup(target)?;
+		let aux = aux.ref_into();
+
+		if Self::nominating(aux).is_some() {
+			// already nominating someone - denominate first.
+			return Err("Cannot nominate when already nominating");
+		}
+
+		// update nominators_for
+		let mut t = Self::nominators_for(&target);
+		t.push(aux.clone());
+		<NominatorsFor<T>>::insert(&target, t);
+
+		// update nominating
+		<Nominating<T>>::insert(aux, &target);
+
+		// Update bondage
+		<Bondage<T>>::insert(aux.ref_into(), T::BlockNumber::max_value());
+
+		Ok(())
+	}
+
+	/// Will panic if called when source isn't currently nominating target.
+	/// Updates Nominating, NominatorsFor and NominationBalance.
+	fn unnominate(aux: &T::PublicAux, target_index: u32) -> Result {
+		let source = aux.ref_into();
+		let target_index = target_index as usize;
+
+		let target = <Nominating<T>>::get(source).ok_or("Account must be nominating")?;
+
+		let mut t = Self::nominators_for(&target);
+		if t.get(target_index) != Some(source) {
+			return Err("Invalid target index")
+		}
+
+		// Ok - all valid.
+
+		// update nominators_for
+		t.swap_remove(target_index);
```

Comment:
Just a question to refine my own understanding: isn't this will invalidate an index of the last nominator?

so for example, there are two calls:

1. `unnominate(0, 0)`
2. `unnominate(1, 1)`

the second call will return an error `Invalid target index` because the first call invalidated the index of the nominator `1`.

Reply:
yes - in that case, the second will be a no-op. shit happens :)